Mathlib/Computability/MyhillNerode.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2024 Google. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+import Mathlib.Computability.DFA
+import Mathlib.Data.Set.Finite
+
+/-!
+# Myhill–Nerode theorem
+
+This file proves the Myhill–Nerode theorem via left quotients.
+
+Given a language `L` and a word `x`, the *left quotient* is the set of suffixes `y` such that
+`x ++ y` is in `L`. The *Myhill–Nerode theorem* shows that each left quotient, in fact, corresponds
+to the state of an automaton that matches `L`, and that `L` is regular if and only if there are
+finitely many such states.
+
+## Implementation notes
+
+Since mathlib doesn't have an `IsRegular` definition yet, we define regularity directly as the
+existence of a DFA.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Syntactic_monoid#Myhill%E2%80%93Nerode_theorem>
+-/
+
+universe u v
+variable {α : Type u} {σ : Type v} (L : Language α)
+
+namespace Language
+
+/-- The *left quotient* of `x` is the set of suffixes `y` such that `x ++ y` is in `L`. -/
+def leftQuotient (x : List α) : Language α := { y | x ++ y ∈ L }
+
+@[simp]
+theorem leftQuotient_nil : L.leftQuotient [] = L := rfl
+
+@[simp]
+theorem leftQuotient_append (x y : List α) :
+    L.leftQuotient (x ++ y) = (L.leftQuotient x).leftQuotient y := by
+  dsimp [leftQuotient, Language]
+  simp_rw [List.append_assoc]
+
+@[simp]
+theorem leftQuotient_mem (x y : List α) : y ∈ L.leftQuotient x ↔ x ++ y ∈ L := Iff.rfl
+
+theorem leftQuotient_accepts (M : DFA α σ) (x : List α) :
+    leftQuotient M.accepts x = M.acceptsFrom (M.eval x) := by
+  ext y
+  rw [DFA.mem_acceptsFrom, DFA.eval, ← DFA.evalFrom_of_append, ← DFA.mem_accepts, leftQuotient_mem]
+
+theorem leftQuotient_accepts' (M : DFA α σ) : leftQuotient M.accepts = M.acceptsFrom ∘ M.eval :=
+  funext <| leftQuotient_accepts M
+
+theorem finite_leftQuotient_of_dfa [Finite σ] (M : DFA α σ) :
```

Comment:
This name doesn't sound right to me.

Reply:
This lemma is now written in terms of `IsRegular`, so it should make more sense.